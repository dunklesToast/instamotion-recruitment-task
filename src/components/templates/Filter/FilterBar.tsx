import React from 'react';
import { Button } from '@atoms/Button/Button';
import { buildFilters } from '@helpers/filterBuilder';
import { RangeSelect } from '@molecules/RangeSelect/RangeSelect';
import { Select } from '@molecules/Select/Select';
import type { Filters, FormikFilters, HTTPFilterBody } from '@type/Filter.type';
import { useFormik } from 'formik';
import * as S from './Filter.styled';

export interface FilterBarProps {
  filters: Filters;
  onSubmit: (filters: HTTPFilterBody) => void;
}

/**
 * Filter Sidebar
 */
export function FilterBar({ filters, onSubmit }: FilterBarProps): JSX.Element {
  const formik = useFormik<FormikFilters>({
    initialValues: {
      make: undefined,
      model: undefined,
      category: undefined,
      gearbox: undefined,
      exteriorColor: undefined,
      fuel: undefined,
      mileage: {
        from: '',
        to: '',
      },
      power: {
        from: '',
        to: '',
      },
      price: {
        from: '',
        to: '',
      },
    },
    onSubmit: (values) => {
      onSubmit(buildFilters(values));
    },
    validate: (values) => {
      onSubmit(buildFilters(values));
    },
    validateOnBlur: true,
  });

  return (
    <S.Filter data-testid='filterbar'>
      <S.Header weight={700}>Filter</S.Header>
      <S.Divider />
      <Select
        options={filters.make}
        name='make'
        localizedName='Marke'
        initField={true}
        onChange={formik.handleChange}
      />
      {formik.values.make && formik.values.make !== '-' && (
        <>
          <S.Divider />
          <Select
            options={filters.model[formik.values.make as unknown as string]}
            name='model'
            localizedName='Modell'
            initField={true}
            onChange={formik.handleChange}
          />
        </>
      )}
      <S.Divider />
      <Select
        options={filters.category}
        name='category'
        localizedName='Kategorie'
        initField={true}
        onChange={formik.handleChange}
      />
      <S.Divider />
      <Select
        options={filters.gearbox}
        name='category'
        localizedName='Schaltung'
        initField={true}
        onChange={formik.handleChange}
      />
      <S.Divider />
      <Select
        options={filters.exteriorColor}
        name='exteriorColor'
        localizedName='Farbe'
        initField={true}
        onChange={formik.handleChange}
      />
      <S.Divider />
      <Select
        options={filters.fuel}
        name='fuel'
        localizedName='Kraftstoff'
        initField={true}
        onChange={formik.handleChange}
      />
      <S.Divider />
      <RangeSelect
        start={formik.values.mileage.from}
        end={formik.values.mileage.to}
        onChange={(e) => {
          formik.setFieldValue('mileage.from', e.start);
          formik.setFieldValue('mileage.to', e.end);
        }}
        name='km Stand'
      />
      <S.Divider />
      <RangeSelect
        start={formik.values.power.from}
        end={formik.values.power.to}
        onChange={(e) => {
          formik.setFieldValue('power.from', e.start);
          formik.setFieldValue('power.to', e.end);
        }}
        name='Leistung'
      />
      <S.Divider />
      <RangeSelect
        start={formik.values.price.from}
        end={formik.values.price.to}
        onChange={(e) => {
          formik.setFieldValue('price.from', e.start);
          formik.setFieldValue('price.to', e.end);
        }}
        name='Preis'
      />
      <S.Divider />
      <Button secondary onClick={() => formik.handleSubmit()} type='submit'>
        Filter anwenden
      </Button>
    </S.Filter>
  );
}
