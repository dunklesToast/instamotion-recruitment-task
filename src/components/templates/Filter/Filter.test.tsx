import React from 'react';
import { render, screen } from '@testing-library/react';
import { FilterBar } from './FilterBar';

test('Filter', () => {
  render(
    <FilterBar
      onSubmit={jest.fn}
      filters={{
        make: [],
        model: {
          test: [],
        },
        gearbox: [],
        carCondition: [],
        category: [],
        doors: [],
        emissionClass: [],
        exteriorColor: [],
        interiorColor: [],
        fuel: [],
        interiorType: [],
        climatisation: [],
        __typename: 'Filters',
      }}
    />
  );
  expect(screen.getByTestId('filterbar')).toBeInTheDocument();
  expect(screen.getByTestId('filterbar')).toMatchSnapshot();
});
