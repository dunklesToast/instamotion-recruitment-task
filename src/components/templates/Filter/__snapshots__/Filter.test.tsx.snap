// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter 1`] = `
.c1 {
  font-family: 'PT Sans',sans-serif;
  font-style: regular;
  font-weight: 700;
}

.c5 {
  font-family: 'PT Sans',sans-serif;
  font-style: regular;
  font-weight: 400;
}

.c8 {
  background-color: #F7B628;
  border: none;
  border-radius: 8px;
  color: black;
  outline: none;
  padding: 10px 28px;
  -webkit-transition: background-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out;
}

.c6 {
  border: 1px black solid;
  border-radius: 4px;
  font-family: 'PT Sans',sans-serif;
  margin-top: 4px;
  padding: 4px;
  width: 80px;
}

.c7 {
  margin: 0 8px;
}

.c4 {
  border: 1px black solid;
  border-radius: 4px;
  font-family: 'PT Sans',sans-serif;
  margin-top: 4px;
  padding: 4px;
}

.c0 {
  background-color: white;
  border-radius: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 16px 16px;
}

.c2 {
  font-size: 24px;
}

.c3 {
  width: 100%;
}

@media (hover:hover) {
  .c8:hover {
    background-color: #cb9712;
    cursor: pointer;
  }
}

@media (max-width:1223px) {
  .c8 {
    padding: 10px 14px;
  }
}

<div
  class="c0"
  data-testid="filterbar"
>
  <p
    class="c1 c2"
  >
    Filter
  </p>
  <hr
    class="c3"
  />
  <p
    class="c1"
  >
    Marke
  </p>
  <select
    class="c4"
    data-testid="select"
    name="make"
  >
    <option
      value=""
    >
      Marke auswählen
    </option>
  </select>
  <hr
    class="c3"
  />
  <p
    class="c1"
  >
    Kategorie
  </p>
  <select
    class="c4"
    data-testid="select"
    name="category"
  >
    <option
      value=""
    >
      Kategorie auswählen
    </option>
  </select>
  <hr
    class="c3"
  />
  <p
    class="c1"
  >
    Schaltung
  </p>
  <select
    class="c4"
    data-testid="select"
    name="category"
  >
    <option
      value=""
    >
      Schaltung auswählen
    </option>
  </select>
  <hr
    class="c3"
  />
  <p
    class="c1"
  >
    Farbe
  </p>
  <select
    class="c4"
    data-testid="select"
    name="exteriorColor"
  >
    <option
      value=""
    >
      Farbe auswählen
    </option>
  </select>
  <hr
    class="c3"
  />
  <p
    class="c1"
  >
    Kraftstoff
  </p>
  <select
    class="c4"
    data-testid="select"
    name="fuel"
  >
    <option
      value=""
    >
      Kraftstoff auswählen
    </option>
  </select>
  <hr
    class="c3"
  />
  <div
    data-testid="range-select"
  >
    <p
      class="c5"
    >
      km Stand
    </p>
    <input
      class="c6"
      placeholder="von"
      type="text"
      value=""
    />
    <span
      class="c7"
    >
      -
    </span>
    <input
      class="c6"
      placeholder="bis"
      type="text"
      value=""
    />
  </div>
  <hr
    class="c3"
  />
  <div
    data-testid="range-select"
  >
    <p
      class="c5"
    >
      Leistung
    </p>
    <input
      class="c6"
      placeholder="von"
      type="text"
      value=""
    />
    <span
      class="c7"
    >
      -
    </span>
    <input
      class="c6"
      placeholder="bis"
      type="text"
      value=""
    />
  </div>
  <hr
    class="c3"
  />
  <div
    data-testid="range-select"
  >
    <p
      class="c5"
    >
      Preis
    </p>
    <input
      class="c6"
      placeholder="von"
      type="text"
      value=""
    />
    <span
      class="c7"
    >
      -
    </span>
    <input
      class="c6"
      placeholder="bis"
      type="text"
      value=""
    />
  </div>
  <hr
    class="c3"
  />
  <button
    class="c8"
    data-testid="button"
    type="submit"
  >
    <p
      class="c5"
    >
      Filter anwenden
    </p>
  </button>
</div>
`;
